[docs] usecase: Accelerating Batch Analytics Part 3 (#3709)
feldera/feldera: docs/use_cases/batch/part3.md

ingested, we prevent any duplicate records.

```json
 {
    "labels": ["lineitem.live"],
    "start_after": "lineitem.historical",
    "transport": {
        "name": "kafka_input",
        "config": {
            "bootstrap.servers": "redpanda:9092",
            "auto.offset.reset": "latest",
            "topics": ["lineitem"]
        }
    },
    "format": {
        "name": "csv",
        "config": {
            "delimiter": "|"
        }
    }
 }
```

### Ensuring Data Consistency with `auto.offset.reset`

In the Kafka connector, we configure **auto.offset.reset** to **latest**. This
ensures that only the latest messages arriving after the consumer starts are
ingested, preventing redundancy.

If **auto.offset.reset** is set to **earliest**, all available messages from the
beginning of the topic, including older historical data, that may have already
been loaded from Delta Lake, will be processed.

:::danger
- Ensure all historical data in Kafka is moved to Delta Lake before Feldera
  starts reading.
- Failure to do so **may result in data loss** due to missing records not
  included in the backfill process.
:::

> ## Recommended Approach: Kafka and Delta Lake

Reviewer: So the previous two sections will be gone, and this section will start with a complete table declaration, followed by:

- Explain connector orchestration with labels
- Then explain the more nuanced issue with Kafka offsets.

  **(0)** and that data has been synced to Delta Lake up to **offset 41**.
  We now want Feldera to start reading from **offset 42**.

```sql
CREATE TABLE LINEITEM (
        L_ORDERKEY    INTEGER NOT NULL,
        L_PARTKEY     INTEGER NOT NULL,
        L_SUPPKEY     INTEGER NOT NULL,
        L_LINENUMBER  INTEGER NOT NULL,
        L_QUANTITY    DECIMAL(15,2) NOT NULL,
        L_EXTENDEDPRICE  DECIMAL(15,2) NOT NULL,
        L_DISCOUNT    DECIMAL(15,2) NOT NULL,
        L_TAX         DECIMAL(15,2) NOT NULL,
        L_RETURNFLAG  CHAR(1) NOT NULL,
        L_LINESTATUS  CHAR(1) NOT NULL,
        L_SHIPDATE    DATE NOT NULL,
        L_COMMITDATE  DATE NOT NULL,
        L_RECEIPTDATE DATE NOT NULL,
        L_SHIPINSTRUCT CHAR(25) NOT NULL,
        L_SHIPMODE     CHAR(10) NOT NULL,
        L_COMMENT      VARCHAR(44) NOT NULL
) WITH (
 'connectors' = '[
  {
    "labels": ["lineitem.historical"],
    "transport": {
      "name": "delta_table_input",
      "config": {
        "uri": "s3://batchtofeldera/lineitem",
        "aws_skip_signature": "true",
        "aws_region": "ap-southeast-2",
        "mode": "snapshot"
      }
    }
  },
  {

>     "labels": ["lineitem.live"],

Reviewer: This label can be removed.

    "start_after": "lineitem.historical",
    "transport": {
        "name": "kafka_input",
        "config": {
            "bootstrap.servers": "localhost:9092",
            "topics": ["lineitem"],
            "start_from": [{
              "topic": "lineitem",
              "partition": 0,
              "offset": 42
            }]
        }
    },
    "format": {
        "name": "json",
        "config": {
            "update_format": "insert_delete",
            "array": false
        }
    }
  }]');
```

### Connector Orchestration with Labels

> Connectors can be labeled according to their role in the pipeline.
> 
> Here, we specify the Delta Lake input connector the label **lineitem.historical**
> as it fetches the historical data from Delta Lake. Similarly, we specify Kafka input
> connector the label **lineitem.live** as it is responsible for getting the
> live data from Kafka.
> 
> Now, we can use these labels to tell one connector to start only after another connector has
> reached the end-of-file (**EOF**). We want the live data to be loaded only after
> the historical data. To do this, in the connector definition of `lineitem.live`, we specify
> that it has to **start_after** the connector with label `lineitem.historical` has completed.
> 

Reviewer: ```suggestion
Connectors can be assigned arbitrary text labels for use in connector orchestration. These labels are typically chosen to reflect the connector's role in the pipeline. Here, we label the Delta Lake input connector  **lineitem.historical**
as it fetches the historical data from Delta Lake. We want the live data to be loaded only after
the historical data. To do this, in the Kafka connector definition, we specify
that it has to **start_after** the connector with label `lineitem.historical` has completed:
```

```json
// lineitem.live
    "start_after": "lineitem.historical",
```

> :::note
> It is also possible to use the Feldera API and Feldera CLI tool `fda` to manually **pause** and **resume**
> these connectors.
> :::

Reviewer: ```suggestion
:::note
It is also possible to use the Feldera API and Feldera CLI tool `fda` to **pause** and **resume**
connectors.
:::
```

Reviewer: Add doc references

### Start Reading Kafka Messages from a Specific Point

> The Kafka input connector can be configured to start reading messages from a specific point.

Reviewer: ```suggestion
The Kafka input connector can be configured to start reading messages from a specific offset in the Kafka topic.
```

> Messages in Kafka are stored at different **offsets** in different **partitions**. A topic can
> have multiple partitions. Use **start_from** key in the Kafka input connector's configuration
> to specify
> 

Reviewer: ```suggestion
```

> To start from a specific point in Kafka, users must specify the **topic**,
> the **partition** within the topic, and the **offset** within the partition. It is possible to
> start at different offsets in different partitions in different topics.

Reviewer: ```suggestion
To this end, the user specifies the **start_from** property with a list of topics,
partitions, and initial offsets within each partition. It is possible to
start at different offsets in different partitions in different topics.
```

> To start from the topic **lineitem**, partition **0** and offset **42**, in the connector configuration, we do:

Reviewer: ```suggestion
To start reading topic **lineitem**, partition **0** from offset **42**, in the connector configuration, we do:
```

```json
// lineitem.live
// config:
            "start_from": [{
              "topic": "lineitem",
              "partition": 0,
              "offset": 42
            }]
```

:::important
In cases with **multiple partitions and multiple topics**, it is necessary to
**specify the offset for each partition and topic**.
:::

## Takeaways

> - Feldera simplifies working with historical and real-time data together by specifying multiple

Reviewer: ```suggestion
- Feldera simplifies working with historical and real-time data within the same pipeine by specifying multiple
```
